components/brave_vpn/utils_win.cc

```diff
@@ -0,0 +1,311 @@
+                     RASEO_ShowDialingProgress;
+  wcscpy_s(entry.szLocalPhoneNumber, 128, hostname);
+  entry.dwfNetProtocols = RASNP_Ip | RASNP_Ipv6;
+  entry.dwFramingProtocol = RASFP_Ppp;
+  wcscpy_s(entry.szDeviceType, 16, RASDT_Vpn);
+  wcscpy_s(entry.szDeviceName, 128, TEXT("WAN Miniport (IKEv2)"));
+  entry.dwType = RASET_Vpn;
+  entry.dwEncryptionType = ET_Optional;
+  entry.dwVpnStrategy = VS_Ikev2Only;
+  entry.dwfOptions2 = RASEO2_DontNegotiateMultilink |
+                      RASEO2_ReconnectIfDropped |
+                      RASEO2_IPv6RemoteDefaultGateway | RASEO2_CacheCredentials;
+  entry.dwRedialCount = 3;
+  entry.dwRedialPause = 60;
+
+  // this maps to "Type of sign-in info" => "User name and password"
+  entry.dwCustomAuthKey = 26;
+
+  DWORD dw_ret = RasSetEntryProperties(DEFAULT_PHONE_BOOK, entry_name, &entry,
+                                       entry.dwSize, NULL, NULL);
+  if (dw_ret != ERROR_SUCCESS) {
+    PrintRasError(dw_ret);
+    return false;
+  }
+
+  if (SetCredentials(entry_name, username, password) != ERROR_SUCCESS) {
+    return false;
+  }
+
+  // Policy needs to be set, otherwise you'll see an error like this in
+  // `eventvwr`:
+  // >> The user DESKTOP - DRCJVG6\brian dialed a connection named BRAVEVPN
+  // which has failed.The error code returned on failure is 13868.
+  //
+  // I've found you can set this manually via PowerShell using the
+  // `Set-VpnConnectionIPsecConfiguration` cmdlet:
+  // https://docs.microsoft.com/en-us/powershell/module/vpnclient/set-vpnconnectionipsecconfiguration?view=windowsserver2019-ps
+  //
+  // I've used the following parameters via PowerShell:
+  // >> AuthenticationTransformConstants: GCMAES256
+  // >> CipherTransformConstants : GCMAES256
+  // >> DHGroup : ECP384
+  // >> IntegrityCheckMethod : SHA256
+  // >> PfsGroup : None
+  // >> EncryptionMethod : GCMAES256
+  //
+  // RAS doesn't expose public methods for editing policy. However, the storage
+  // is just an INI format file:
+  // `%APPDATA%\Microsoft\Network\Connections\Pbk\rasphone.pbk`
+  //
+  // The variable being set in this file is similar to the structure
+  // `ROUTER_CUSTOM_IKEv2_POLICY0` which was part of MPR (Multiprotocol
+  // Routing). The DWORDs are written out byte by byte in 02d format as
+  // `CustomIPSecPolicies` and `NumCustomPolicy` is always being set to 1.
+  //
+  // NOTE: *This IKEv2 implementation (due to policy) might only be supported on
+  // Windows 8 and above; we need to check that.*
+  //
+
+  constexpr wchar_t kNumCustomPolicy[] = L"1";
+  constexpr wchar_t kCustomIPSecPolicies[] =
+      L"020000000600000005000000080000000500000000000000";
+  std::wstring phone_book_path = GetPhonebookPath();
+  if (phone_book_path.empty())
+    return false;
+
+  // https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-writeprivateprofilestringw
+  BOOL wrote_entry =
+      WritePrivateProfileString(entry_name, L"NumCustomPolicy",
```

Comment:
Are we sure that's the right way to use this API? There is a deprecation notice at the top of https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-writeprivateprofilestringw:

> Note  This function is provided only for compatibility with 16-bit versions of Windows. Applications should store initialization information in the registry.

Reply:
@fmarier I assume above comments were written to recommend to use registry for storing application init data instead of `.ini` file because this `WritePrivateProfileString` api is used to edit ini file format.
We are using this api just to edit phonebook file(.pbk) because this phonebook file is `ini` format.
IMO, I think it's safe to use this api just for editing phonebook file(ini format). cc @bsclifton

Reply:
Good question @fmarier - I think the risk is acceptable. This is definitely not a preferred API by any means... but if this gets removed we can always revisit (ex: provide a polyfill or replace with a `sed` equivalent). There is a [UWP implementation for VPN methods](https://docs.microsoft.com/en-us/uwp/api/windows.networking.vpn?view=winrt-20348) which I'm sure Microsoft would prefer be used but the integration is a lot heavier

Reply:
> I think the risk is acceptable. This is definitely not a preferred API by any means... but if this gets removed we can always revisit (ex: provide a polyfill or replace with a sed equivalent).

If there's one company that knows how to maintain software/APIs for the long term, it's Microsoft, so if you're comfortable relying on what sounds like pre-Win95 API, then :+1: